src/xenia/base/memory.cc

```diff
@@ -215,7 +219,102 @@ void copy_and_swap_16_in_32_unaligned(void* dest_ptr, const void* src_ptr,
     dest[i] = (src[i] >> 16) | (src[i] << 16);
   }
 }
+
+#elif XE_ARCH_ARM64
+
+// Although NEON offers vector rev instructions (like vrev32q_u8), they are
+// slower in benchmarks. Also, using uint8x16xN_t wasn't any faster in the
+// benchmarks, hence we use just use one SIMD register to minimize residual
+// processing.
+
+void copy_and_swap_16_aligned(void* dst_ptr, const void* src_ptr,
+                              size_t count) {
+  copy_and_swap_16_unaligned(dst_ptr, src_ptr, count);
+}
+
+void copy_and_swap_16_unaligned(void* dst_ptr, const void* src_ptr,
+                                size_t count) {
+  auto dst = reinterpret_cast<uint16_t*>(dst_ptr);
+  auto src = reinterpret_cast<const uint16_t*>(src_ptr);
+
+  const uint8x16_t tbl_idx = vcombine_u8(vcreate_u8(0x0607040502030001),
+                                         vcreate_u8(0x0E0F0C0D0A0B0809));
+
+  int64_t i;
```

Comment:
Signedness is not needed here — `size_t` is the preferred type for indices in C++, and the comparison can be changed to `i + 7 < count`.

Reply:
I did that on purpose to simplify the loop condition check. (count + x) can be precomputed and I think the compiler can't make that optimization (from (i + 8 <= count) to (i <= count + 7)) itself because it requires typecasting size_t -> int64_t.

Reply:
Is the generated code good if the loop counter is changed from the element index to the vector index?

Reply:
didn't check that explicitly. we need element index for residual handling later anyways. 
I might rework it  so that it downcounts `count` like:

```c
while (count >= 1234) {
  // vector stuff , dst, src
  count -= 1234;
  dst += 1234 * sizeof();
  src += 1234 * sizeof();
}
while (count > 0) {
 // single bswap
 count--;
 dst += sizeof();
 src += sizeof();
}
```

That might be easier for a risc arch which doesn't have the crazy memory stride calculations (`LEA`) that x86 does.

Reply:
Sooo... armv8-a clang 11.0.1  -O2

inst_count
65              using size_t and swapping the condition like you suggested
60              current int64_t solution
53              solution from last comment